import numpy as np
reps = np.arange(8)
hg1_sim_reps = np.arange(8)
snp_stat_reps = np.arange(10)


# ===== rules for producing neutral simulation results ======
rule all_neutral_wv_sim:
	input:
		expand("results/neutral_sims/{sim}/ancestry_master.txt", sim = ['equilibrium', 'bottleneck'])

rule neutral_sims: # this runs simulations for neutral admixture pulse: (1) demographic equilibrium and (2) early bottleneck
	input:
	output:
		"results/neutral_sims/{sim}/replicate{rep}.trees" #trees file includes remembered individuals from various gens
	shell:
		"slim -d seed={wildcards.rep} -d N=10000 code/{wildcards.sim}.slim"

rule ancestry_neutral_sims: # calculates ancestry fractions for the neutral sims
	input:
		"results/neutral_sims/{sim}/replicate{rep}.trees"
	output:
		expand("results/neutral_sims/{sim}/replicate{rep}_{data}.txt", data=['haps','frqs','ancestry'])
	shell:
		"python code/ancestry_haps_frqs.py {input} 10000 10"


rule simulation_swordtail:
	input:
	output:
		temp(expand("results/swordtail_neutral/swordtail_replicate{{rep}}_{gen}_N{{N}}.trees", gen = gens_swordtail))
	shell:
		"slim -d seed={wildcards.rep} -d N={wildcards.N} code/swordtail_neutral.slim"

rule avg_ancestry_calc_swordtail:
	input:
		"results/swordtail_neutral/swordtail_{rep_gen_N}.trees"
	output:
		temp("results/swordtail_neutral/swordtail_{rep_gen_N}_ancestry.txt")
	shell:
		"python3 code/ancestry.py {input} > {output}"
	
# ===== rules for simulations of human chromosome 1  =======

rule human_chrom_sims:
	input:
	output:
		"results/human_chr1{sim}/replicate{rep}.trees"
	shell:
		"slim -d seed={wildcards.rep} -d N=10000 code/human_chr1{wildcards.sim}.slim"

rule all_human_chrom_sims:
	input:
		expand("results/human_chr1{sim}/replicate{rep}.trees", sim = ["neutral","sel1-1000"], rep = np.arange(10))

rule output_human_chrom_ancestry:
	input:
		"results/human_chr1{sim}/replicate{rep}.trees"
	output:
		"results/human_chr1{sim}/replicate{rep}_frqs.txt"
	shell:
		"python code/ancestry-multigen.py {input} 3 7 12 52 102 502 1002 > {output}"

rule combine_human_chrom_ancestry:
	input:
		expand("results/human_chr1{{sim}}/replicate{rep}_frqs.txt", rep=np.arange(10))
	output:
		"results/human_chr1{sim}/all_frq.txt"
	shell:
		"cat {input} > {output}"

rule all_human_chrom_ancestry:
	input:
		expand("results/human_chr1{sim}/all_frq.txt", sim=["_neutral","_sel1-1000", "_sel1-10","_sel100-200"])

# ===== rules for simulations with periodic recombination =====

rule periodic_recomb_sims:
	input:
	output:
		"results/periodic_recomb_{sim}/replicate{rep}.trees" 
	shell:
		"slim -d seed={wildcards.rep} -d N=10000 code/periodic_recomb_{wildcards.sim}.slim"	

rule periodic_recomb_all_sims:
	input:
		expand('results/periodic_recomb_{sim}/replicate{rep}.trees', sim = ['sel_constant', 'sel_reverses'], rep=np.arange(10))

rule periodic_recomb_ancestry:
	input:
		"results/periodic_recomb_{sim}/replicate{rep}.trees"
	output:
		"results/periodic_recomb_{sim}/replicate{rep}_ancestry.txt"
	shell:
		"python code/derived-allele-freq-general.py {input} 1 2 3 4 5 10 25 50 100 250 500 1000 > {output}"

rule periodic_recomb_combine_ancestry:
	input:
		expand("results/periodic_recomb_{{sim}}/replicate{rep}_ancestry.txt", rep = np.arange(10))
	output:
		"results/periodic_recomb_{sim}/ancestry_master.txt"
	shell:
		"cat {input} > {output}"

rule all_periodic_recomb_ancestry:
	input:
		expand("results/periodic_recomb_{sim}/ancestry_master.txt", sim = ['sel_constant', 'sel_reverses'])

# ===== allele frq based statistic sims =====

rule snp_stat_const_rec_neutral_sim:
	input:
	output:
		"results/snp_stat_const_rec/replicate{rep}.trees"
	shell:
		"slim -d seed={wildcards.rep} -d N=100000 code/snp_stat_const_rec.slim"

rule make_variable_rec_map:
	input:
	output:
		"variable_rec_map_slim.txt",
		"variable_rec_map_msprime.txt"
	shell:
		"Rscript --vanilla code/make_variable_rec_map.R"

rule snp_stat_variable_rec_sel_sim:
	input:
		"variable_rec_map_slim.txt"
	output:
		"results/snp_stat_variable_rec_sel/replicate{rep}.trees"
	shell:
		"slim -d seed={wildcards.rep} -d N=10000 code/snp_stat_variable_rec_sel.slim"

rule snp_stat_variable_rec_neutral_sim:
	input:
		"variable_rec_map_slim.txt"
	output:
		"results/snp_stat_variable_rec_neutral/replicate{rep}.trees"
	shell:
		"slim -d seed={wildcards.rep} -d N=10000 code/snp_stat_variable_rec_neutral.slim"

rule all_snp_stat_variable_rec_sims:
	input:
		expand("results/snp_stat_variable_rec_{sim}/replicate{rep}.trees", sim = ['sel', 'neutral'], rep = np.arange(10))

rule snp_stat_const_rec_ancestry:
	input:
		"results/snp_stat_const_rec/replicate{rep}.trees"
	output:
		expand("results/snp_stat_const_rec/replicate{{rep}}_{data}.txt", data=["haps","frqs","ancestry"])
	shell:
		"python code/snp_stat_const_rec.py {input} 10000 10" #2nd argument is diploid size, 3rd argument is number of haplotypes to output

rule all_snp_stat_const_rec_ancestry:
	input:
		expand('results/snp_stat_const_rec/replicate{rep}_{data}.txt', data=['haps', 'frqs', 'ancestry'], rep=np.arange(10))

rule snp_stat_variable_rec_ancestry:
	input:
		trees="results/snp_stat_variable_rec_{sim}/replicate{rep}.trees",
		recmap="variable_rec_map_msprime.txt"
	output:
		"results/snp_stat_variable_rec_{sim}/replicate{rep}_frqs.txt"
	shell:
		"python code/snp_stat_variable_rec.py {input.trees} 10000 10"

rule snp_stat_const_rec_wavelets:
	input:
		frqs="results/snp_stat_const_rec/replicate{rep}_frqs.txt", 
		haps="results/snp_stat_const_rec/replicate{rep}_haps.txt"
	output:
		"results/snp_stat_const_rec/replicate{rep}_wv_results.RData"
	shell: # first argument to R script is the number of haplotypes output, should match the output from the previous rule that outputs ancestry
		"""
		Rscript --vanilla code/admix_snp_stat.R 10 {input.haps} {input.frqs}
		"""

rule snp_stat_variable_rec_wavelets:
	input:
		frqs="results/snp_stat_variable_rec_{sim}/replicate{rep}_frqs.txt",
		recmap="variable_rec_map_slim.txt"
	output:
		"results/snp_stat_variable_rec_{sim}/replicate{rep}_wv_results.txt"
	shell:
		"Rscript --vanilla code/snp_stat_variable_rec.R {input.frqs} {input.recmap} {wildcards.sim} > {output}"

rule snp_stat_variable_rec_all_wavelet_results:
	input:
		expand("results/snp_stat_variable_rec_{sim}/replicate{rep}_wv_results.txt",sim = ['sel', 'neutral'], rep = np.arange(10))
	output:
		"results/snp_stat_variable_rec_sel/all_wv_results_sel_v_neutral.txt"
	shell:
		"cat {input} > {output}"

# ========


rule dwt_on_haps:
	input:
		"results/equilibrium_n500/{rep_gen}_allHaps.txt"
	output:
		"results/equilibrium_n500/{rep_gen}_variances-table.txt"
	shell:
		"Rscript --vanilla code/all-variances-table.R {input} {wildcards.rep_gen} {output}"


#rule aggregate_mean_ancestry:
#	input:
#		input = expand("results/{{sim}}/replicate{rep}_{gen}_ancestry.txt", rep=reps,gen=gens)
#	output:
#		output="results/{sim}/ancestry_master.txt"
#	shell:
#		"""
#        cat {input} > results/{wildcards.sim}/ancestry_master.txt
#        """

rule aggregate_mean_ancestry_swordtail:
	input:
		expand("results/swordtail_neutral/swordtail_replicate{rep}_{gen}_N{N}_ancestry.txt", rep=reps, gen=gens_swordtail,N=swordtail_N)
	output:
		output="results/swordtail_neutral/ancestry_master.txt"
	shell:
		"""
        cat {input} > {output}
        """

rule aggregate_variance_tables:
	input:
		expand("results/equilibrium_n500/replicate{rep}_{gen}_variances-table.txt", rep=reps,gen=gens)
	output:
		"results/equilibrium_n500/all-variances-master.txt"
	shell:
		"Rscript --vanilla code/rbind_variance_tables.R {output} {input}"
